tests/native_api_tests/test_api_model.cpp

```diff
@@ -83,34 +93,49 @@ TEST_CASE("API Model") {
     int8_t load_type = 2;
     double load_p_specified = 0.0;
     double load_q_specified = 500.0;
-    Buffer load_buffer{PGM_def_input_sym_load, 1};
+    Buffer const load_buffer{PGM_def_input_sym_load, 1};
     load_buffer.set_value(PGM_def_input_sym_load_id, &load_id, -1);
     load_buffer.set_value(PGM_def_input_sym_load_node, &load_node, -1);
     load_buffer.set_value(PGM_def_input_sym_load_status, &load_status, -1);
     load_buffer.set_value(PGM_def_input_sym_load_type, &load_type, -1);
     load_buffer.set_value(PGM_def_input_sym_load_p_specified, &load_p_specified, -1);
     load_buffer.set_value(PGM_def_input_sym_load_q_specified, &load_q_specified, -1);
 
-    // add buffers
-    input_dataset.add_buffer("node", 1, 1, nullptr, node_buffer);
+    // add buffers - row
     input_dataset.add_buffer("sym_load", 1, 1, nullptr, load_buffer);
     input_dataset.add_buffer("source", 1, 1, nullptr, source_buffer);
 
+    // add buffers - columnar
+    input_dataset.add_buffer("node", 2, 2, nullptr, nullptr);
+    input_dataset.add_attribute_buffer("node", "id", node_id.data());
+    input_dataset.add_attribute_buffer("node", "u_rated", node_u_rated.data());
+    input_dataset.add_buffer("line", 2, 2, nullptr, nullptr);
+    input_dataset.add_attribute_buffer("line", "id", line_id.data());
+    input_dataset.add_attribute_buffer("line", "from_node", line_from_node.data());
+    input_dataset.add_attribute_buffer("line", "to_node", line_to_node.data());
+    input_dataset.add_attribute_buffer("line", "from_status", line_from_status.data());
+    input_dataset.add_attribute_buffer("line", "to_status", line_to_status.data());
+
     // output data
-    Buffer node_output{PGM_def_sym_output_node, 1};
+    Buffer node_output{PGM_def_sym_output_node, 2};
     node_output.set_nan();
-    DatasetMutable single_output_dataset{"sym_output", 0, 1};
-    single_output_dataset.add_buffer("node", 1, 1, nullptr, node_output);
-    Buffer node_batch_output{PGM_def_sym_output_node, 2};
+    DatasetMutable const single_output_dataset{"sym_output", 0, 1};
+    single_output_dataset.add_buffer("node", 2, 2, nullptr, node_output);
+    Buffer node_batch_output{PGM_def_sym_output_node, 4};
     node_batch_output.set_nan();
-    DatasetMutable batch_output_dataset{"sym_output", 1, 2};
-    batch_output_dataset.add_buffer("node", 1, 2, nullptr, node_batch_output);
+    DatasetMutable const batch_output_dataset{"sym_output", 1, 2};
+    batch_output_dataset.add_buffer("node", 2, 4, nullptr, node_batch_output);
```

Comment:
@TonyXiang8787 in the core, the downstream function calls after `add_buffer` for output do not verify whether the shapes match between the output buffer and the actual result. For example, if we run batch size 4 with 10 nodes provided in columnar format, we would be expecting a 4x10 or a 1x40 holder for the result. However, in reality, we could also provide e.g., 1x10 to it and only the first 10 results will be filled to the holder. Is this behavior intentional?

Reply:
I guess we can add extra checks in the main model and throw errors if the size does not match.

However, this does not change the fact that users can still provide crazy C-API input and we cannot do anything about it. The user can create a buffer of size `10` and tell the C-API dataset that the size is `100`. We do not have any way to verify that.

As with any C-API, we can only document the expected input requirements, and if user does not specify, it is UB.

Reply:
Created an issue: https://github.com/PowerGridModel/power-grid-model/issues/722